### PR TITLE
OSHMEM/SEGMENT-REGISTRATION: added segment filtering

### PR DESCRIPTION
**Pull request by hoopoepg**
- all file-mapped segments are filtered by start-process
  segments only
- segments from another modules are ignored

Signed-off-by: Sergey Oblomov <sergeyo@nvidia.com>

### CODE REVIEW

File: oshmem/mca/memheap/base/memheap_base_static.c [hunk 1]
```diff
@@ -127,7 +130,50 @@ static int _check_address(struct map_segment_desc *seg)
...
+
+            init = 1;
+        }
+
+        pthread_mutex_unlock(&mutex);
+    }
+
+    /* for file-mapped segments allow segments from start process only */
+    /* do not use segment inode - it is not same as file inode */
+    res = stat(seg->pathname, &st);
```
Comment (yosefe) [inline oshmem/mca/memheap/base/memheap_base_static.c:165]: maybe just compare pathname ( realpath(proc/self/exe)==realpath(seg->path) ) ?

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

File: oshmem/mca/memheap/base/memheap_base_static.c [hunk 2]
```diff
@@ -127,7 +130,47 @@ static int _check_address(struct map_segment_desc *seg)
...
+            }
+
+            init = 1;
+        }
+
+        pthread_mutex_unlock(&mutex);
+    }
+
+    /* for file-mapped segments allow segments from start process only */
+    path = realpath(seg->pathname, module_path);
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply oshmem/mca/memheap/base/memheap_base_static.c:161]: yes, there is inode value in segment, but this is not same inode as returned by stat function

File: oshmem/mca/memheap/base/memheap_base_static.c [hunk 3]
```diff
@@ -127,7 +130,47 @@ static int _check_address(struct map_segment_desc *seg)
...
+
+    if (0 == seg->inode) {
+        /* segment is not mapped to file, allow sharing */
+        return OSHMEM_SUCCESS;
+    }
+
+    if (0 == init) {
+        pthread_mutex_lock(&mutex);
+        if (0 == init) {
+            path = realpath(proc_self_exe, exe_path);
```
Comment (yosefe) [inline oshmem/mca/memheap/base/memheap_base_static.c:148]: imo, better not cache it: it increases binary size and almost does not improve perf. we call realpath once anyway for each seg.

Comment (hoopoepg) [reply oshmem/mca/memheap/base/memheap_base_static.c]: ok, caching is removed

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
bot:ompi:retest

**Comment by ggouaillardet** [comment]
minor typo in the commit message

`s/serment/segment/g`

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Pull request merged**